ipn/ipnserver/actor.go

```diff
@@ -0,0 +1,192 @@
+	if err != nil {
+		return nil, err
+	}
+	return &actor{logf: logf, ci: ci, isLocalSystem: connIsLocalSystem(ci)}, nil
+}
+
+// IsLocalSystem implements [ipnauth.Actor].
+func (a *actor) IsLocalSystem() bool {
+	return a.isLocalSystem
+}
+
+// IsLocalAdmin implements [ipnauth.Actor].
+func (a *actor) IsLocalAdmin(operatorUID string) bool {
+	return a.isLocalSystem || connIsLocalAdmin(a.logf, a.ci, operatorUID)
+}
+
+// UserID implements [ipnauth.Actor].
+func (a *actor) UserID() ipn.WindowsUserID {
+	return a.ci.WindowsUserID()
+}
+
+func (a *actor) Pid() int {
+	return a.ci.Pid()
+}
+
+// Username implements [ipnauth.Actor].
+func (a *actor) Username() (string, error) {
+	if a.ci == nil {
+		a.logf("[unexpected] missing ConnIdentity in ipnserver.actor")
+		return "", errors.New("missing ConnIdentity")
+	}
+	switch runtime.GOOS {
+	case "windows":
+		tok, err := a.ci.WindowsToken()
+		if err != nil {
+			return "", fmt.Errorf("get windows token: %w", err)
+		}
+		defer tok.Close()
+		return tok.Username()
+	case "darwin", "linux":
+		uid, ok := a.ci.Creds().UserID()
+		if !ok {
+			return "", errors.New("missing user ID")
+		}
+		u, err := osuser.LookupByUID(uid)
+		if err != nil {
+			return "", fmt.Errorf("lookup user: %w", err)
+		}
+		return u.Username, nil
+	default:
+		return "", errors.New("unsupported OS")
+	}
+}
+
+// actorContextKey is the http.Request.Context's context.Value key for either
+// a pointer to [actor] or an error.
+type actorContextKey struct{}
+
+var errNoActor = errors.New("connection actor not available")
+
+// contextWithActor returns a new context that carries the identity of the actor
+// owning the other end of the [net.Conn]. It can be retrieved with [actorFromContext].
+func contextWithActor(ctx context.Context, logf logger.Logf, c net.Conn) context.Context {
+	actor, err := newActor(logf, c)
+	if err != nil {
+		return context.WithValue(ctx, actorContextKey{}, err)
+	}
+	return context.WithValue(ctx, actorContextKey{}, actor)
+}
+
+// actorFromContext returns an [connIdentity] associated with ctx,
+// or an error if the context does not carry an actor's identity.
+func actorFromContext(ctx context.Context) (*actor, error) {
+	switch v := ctx.Value(actorContextKey{}).(type) {
+	case *actor:
+		return v, nil
+	case error:
+		return nil, v
+	case nil:
+		return nil, errNoActor
+	default:
+		panic("unreachable")
+	}
+}
+
+func connIsLocalSystem(ci *ipnauth.ConnIdentity) bool {
+	token, err := ci.WindowsToken()
+	return err == nil && token.IsLocalSystem()
+}
+
+// connIsLocalAdmin reports whether the connected client has administrative
+// access to the local machine, for whatever that means with respect to the
+// current OS.
+//
+// This is useful because tailscaled itself always runs with elevated rights:
+// we want to avoid privilege escalation for certain mutative operations.
+func connIsLocalAdmin(logf logger.Logf, ci *ipnauth.ConnIdentity, operatorUID string) bool {
```

Comment:
This has been moved from https://github.com/tailscale/tailscale/commit/b78df4d48ab81e211650d33fcbdaecb7b6f6b87d/ipn/localapi/localapi.go#L1075